File: src/vta/BaseVTA.py
from m5.params import *
from m5.proxy import *
from m5.SimObject import SimObject
from m5.objects.Buffer import *
from m5.objects.CommandQueue import *
from m5.objects.ComputeModule import *
from m5.objects.DependencyQueue import *
from m5.objects.InstructionFetchModule import *
from m5.objects.LoadModule import *
from m5.objects.StoreModule import *


class BaseVTA(SimObject):
    type = "BaseVTA"
    cxx_header = "vta/base_vta.hh"
    cxx_class = "gem5::BaseVTA"

    instruction_port = RequestPort("instruction port")
    micro_op_port = RequestPort("micro op port")
    load_data_port = RequestPort("load data port")
    compute_data_port = RequestPort("compute data port")
    store_data_port = RequestPort("store data port")

    system = Param.System(Parent.any, "system object")

    # VTA configs
    # reference at https://tvm.apache.org/docs/topic/vta/dev/hardware.html

    load_command_queue = CommandQueue()
    compute_command_queue = CommandQueue()
    store_command_queue = CommandQueue()

    compute_to_store_queue = DependencyQueue()
    store_to_compute_queue = DependencyQueue()
    compute_to_load_queue = DependencyQueue()
    load_to_compute_queue = DependencyQueue()

    input_buffer = Buffer()
    weight_buffer = Buffer()
    output_buffer = Buffer()

    instruction_fetch_module = InstructionFetchModule()
    instruction_fetch_module.load_command_queue = load_command_queue
    instruction_fetch_module.compute_command_queue = compute_command_queue
    instruction_fetch_module.store_command_queue = store_command_queue
    instruction_fetch_module = Param.InstructionFetchModule(
        instruction_fetch_module,
        "instruction fetch module",
    )

    load_module = LoadModule()
    load_module.load_command_queue = load_command_queue
    load_module.load_to_compute_queue = load_to_compute_queue
    load_module.compute_to_load_queue = compute_to_load_queue
    load_module.input_buffer = input_buffer
    load_module.weight_buffer = weight_buffer
    load_module = Param.LoadModule(load_module, "load module")

    compute_module = ComputeModule()
    compute_module.compute_command_queue = compute_command_queue
    compute_module.load_to_compute_queue = load_to_compute_queue
    compute_module.compute_to_load_queue = compute_to_load_queue
    compute_module.compute_to_store_queue = compute_to_store_queue
    compute_module.store_to_compute_queue = store_to_compute_queue
    compute_module.input_buffer = input_buffer
    compute_module.weight_buffer = weight_buffer
    compute_module.output_buffer = output_buffer
    compute_module = Param.ComputeModule(compute_module, "compute module")

    store_module = StoreModule()
    store_module.store_command_queue = store_command_queue
    store_module.compute_to_store_queue = compute_to_store_queue
    store_module.store_to_compute_queue = store_to_compute_queue
    store_module.output_buffer = output_buffer
    store_module = Param.StoreModule(store_module, "store module")

File: src/vta/base_vta.cc
#include "vta/base_vta.hh"

#include "base/trace.hh"
#include "debug/BaseVTAFlag.hh"
#include "vta/buffer.hh"
#include "vta/compute_module.hh"
#include "vta/data_queue.hh"
#include "vta/instruction_fetch_module.hh"
#include "vta/instruction_queue.hh"
#include "vta/load_module.hh"
#include "vta/store_module.hh"

namespace gem5
{
BaseVTA::BaseVTA(const Params &params) :
    SimObject(params),
    begin_event([this] { processBeginEvent(); }, name() + ".beginEvent"),
    inputBuffer(new Buffer(this)),
    weightBuffer(new Buffer(this)),
    outputBuffer(new Buffer(this)),
    computeToStoreQueue(new DataQueue(this)),
    storeToComputeQueue(new DataQueue(this)),
    computeToLoadQueue(new DataQueue(this)),
    loadToComputeQueue(new DataQueue(this)),
    loadQueue(new InstructionQueue(this)),
    computeCommandQueue(new InstructionQueue(this)),
    storeQueue(new InstructionQueue(this)),
    instructionFetchModule(new InstructionFetchModule(this)),
    computeModule(new ComputeModule(this)),
    loadModule(new LoadModule(this)),
    storeModule(new StoreModule(this))
{
    DPRINTF(BaseVTAFlag, "===========================================\n");
    DPRINTF(BaseVTAFlag, "initializing gem5-vta ... user configuration "
                         "parameters are as follows: \n");
    DPRINTF(BaseVTAFlag, "===========================================\n");

    // connection
    // instruction fetch module
    instructionFetchModule->loadCommandQueue = loadQueue;
    instructionFetchModule->computeCommandQueue = computeCommandQueue;
    instructionFetchModule->storeCommandQueue = storeQueue;

    // computer module
    computeModule->computeCommandQueue = computeCommandQueue;
    computeModule->loadToComputeQueue = loadToComputeQueue;
    computeModule->computeToLoadQueue = computeToLoadQueue;
    computeModule->storeToComputeQueue = storeToComputeQueue;
    computeModule->computeToStoreQueue = computeToStoreQueue;
    computeModule->inputBuffer = inputBuffer;
    computeModule->weightBuffer = weightBuffer;
    computeModule->outputBuffer = outputBuffer;

    // load module
    loadModule->loadCommandQueue = loadQueue;
    loadModule->computeToLoadQueue = computeToLoadQueue;
    loadModule->loadToComputeQueue = loadToComputeQueue;
    loadModule->inputBuffer = inputBuffer;
    loadModule->weightBuffer = weightBuffer;

    // store module
    storeModule->storeCommandQueue = storeQueue;
    storeModule->computeToStoreQueue = computeToStoreQueue;
    storeModule->storeToComputeQueue = storeToComputeQueue;
    storeModule->outputBuffer = outputBuffer;
}

void
BaseVTA::startup()
{
    // DPRINTF(BaseVTAFlag, "%s: start up in BaseVTA.\n", __func__);
    // schedule(begin_event, 100);
}

void
BaseVTA::processBeginEvent()
{
    // DPRINTF(BaseVTAFlag, "%s: process event up in BaseVTA.\n", __func__);
}
} // namespace gem5
